File: src/main/resources/static/R/Histogram.R
# OpenVA - Open software platform for visual analytics
#
# Copyright (c) 2018, VTT Technical Research Centre of Finland Ltd
# All rights reserved.
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 
# 1) Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
# 2) Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
# 3) Neither the name of the VTT Technical Research Centre of Finland nor the
# names of its contributors may be used to endorse or promote products
# derived from this software without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

# Author: Paula Järvinen, Pekka Siltanen
                                                                                                                                               
library(RPostgreSQL)

args <- commandArgs(trailingOnly = F)
scriptPath <- dirname(sub("--file=","",args[grep("--file",args)]))
source(paste(scriptPath, "common.R", sep="/"))

connectDB <- function()
{
	psql <- dbDriver("PostgreSQL")
	dbcon <- dbConnect(psql, host="<host>", port=<port>, dbname="<dbname>",user="<user>",pass="<password>")
	return(dbcon)
}


deploy_histogram=function(variableid, oitype,oiids=NULL, starttime=NULL, endtime=NULL,plot_time_unit,dbcon,imagetype)
{  
    on.exit(dbDisconnect(dbcon))
    my_meta=getMetadata(dbcon,variableid)
    
    print(paste(variableid,oiids,oitype,starttime,endtime,plot_time_unit))
    my_frame=getVariableValues_partition_aggregated_by_time(dbcon,variableid, oiids,oitype,starttime,endtime,fast=FALSE, plot_time_unit,row_limit=NULL) 
print("getVariableValues_partition_aggregated_by_time OK")
    n=nrow(my_frame)

    if(n==0) {
        stop (paste("OpenVA warning: no data found",   my_meta$report_title, sep=" "))
    }
    
    #time unit transformations 
     #print(paste("time unit transformations starts",Sys.time()))
        operation=my_meta$plottype
        oiids_temp=paste("(",oiids,")",sep="")
        my_ois=getOIs(dbcon,oiids_temp,oitype=my_meta$oitype_title)
        
        #return_list=create_ts_object_quantitative(my_frame, plot_time_unit,variableid,meta=my_meta) #saa tulla warning
     #print("end")
        zoo_frame=zoo(my_frame$measurement_value,my_frame$measurement_time)
        timeunit=paste(plot_time_unit,"s",sep="")
        plot_zoo=period.apply(zoo_frame, endpoints(zoo_frame, timeunit), operation)
        plot_zoo <- na.omit(plot_zoo)
        zoo_frame=fortify.zoo(plot_zoo)
        colnames(zoo_frame)=c("measurement_time","measurement_value")
        plot_frame=zoo_frame

        #plot_frame=my_frame 
        plot_time_unit_title=plot_time_unit 
        total_sum=sum(my_frame$measurement_value)
        rm(my_frame)
        gc()
        #print(head(plot_frame))
    #title preparations
        starttime=min(plot_frame$measurement_time)
        endtime=max(plot_frame$measurement_time)

        ois_title=paste(my_ois[,c("report_title")],collapse = ',')
        yLabel = "observations"
        xLabel = paste(my_meta$numunit, plot_time_unit_title)
        #print(xLabel)
  #plot histogram
        histData <- plothist(plot_frame$measurement_value,my_meta, ois_title=ois_title,starttime=starttime,endtime=endtime, yLabel,xLabel,total_sum,total_n=n)

   #return plot data
        return_matrix=as.matrix(plot_frame[1,c("measurement_value","measurement_time")])
        if (my_meta$quanttype=='integer') {
            decimals=0 
        } else {
            decimals=3
        } 
        mean=round(mean(plot_frame$measurement_value),decimals)
        sum=round(sum(plot_frame$measurement_value),decimals)
        std=round(sd(plot_frame$measurement_value),decimals)
        min=round(min(plot_frame$measurement_value),decimals)
        max=round(max(plot_frame$measurement_value),decimals)
        sub=paste("mean=", mean,"sum=",sum,"std=",std,"min=",min,"max=",max)
        main=paste(my_meta$report_title,"\n",ois_title,"\nn=",n,"\n",starttime,"-",endtime)
        return_list <- list(histData,main,xLabel,yLabel,sub,my_meta$lowerlimit,my_meta$upperlimit,mean,sum,std,min,max,my_meta$alarm_lowerlimit,my_meta$alarm_upperlimit,imagetype)
        names(return_list) <-c("matrix","main_title", "x_title","y_title","sub_title","lowerlimit","upperlimit","mean","sum","std","min","max","alarm_lowerlimit","alarm_upperlimit","imagetype")
            rm(plot_frame)
        gc()
    return(return_list)
}


plothist=function(variable_values,meta,ois_title,starttime=NULL,endtime=NULL, yLabel,xLabel,total_sum,total_n)
{  
    barcol="grey"
    n=length(variable_values)
    #if all equal? 
    if(isTRUE(all.equal( max(variable_values) ,min(variable_values)) )) {
    
        plot(x=variable_values, y = seq(1:length(variable_values)), type='l', main=paste(meta$report_title,"\n",ois_title,"\nn=",n,"\n",starttime,"-",endtime),
        sub=paste( 
                 "mean=",round(mean(variable_values),1),
                 "sum=",round(total_sum,1),
                 "std=",round(sd(variable_values),1),
                 "min=",round(min(variable_values),1),
                 "max=",round(max(variable_values),1)),
                  freq=TRUE,col=barcol, ylab=yLabel,xlab=xLabel        )
        h <- variable_values   

    } else {
        #normal histogram 
        #set limits for the image 
        #print("limits") 
        lowerlimit=meta$lowerlimit
        if (is.na(lowerlimit)) 
            lowerlimit=min(variable_values)
        lower_alarmlimit=meta$lower_alarmlimit
        if (is.null(lower_alarmlimit)) 
            lower_alarmlimit=min(variable_values)
        xlim_lowend=min(min(variable_values),lowerlimit, lower_alarmlimit)
        xlim_lowend=xlim_lowend+(xlim_lowend/10)
        upperlimit=meta$upperlimit
        if (is.na(upperlimit)) 
            upperlimit=max(variable_values)
        upper_alarmlimit=meta$upper_alarmlimit
        if (is.null(upper_alarmlimit)) 
            upper_alarmlimit=max(variable_values)
        xlim_uppend=max(max(variable_values),upperlimit, upper_alarmlimit)
        xlim_uppend=xlim_uppend+(xlim_uppend/10)

        #plot    
        h <-hist(variable_values
        ,main=paste(meta$report_title,"\n",ois_title,"\nN=",n,"n=",total_n,"\n",starttime,"-",endtime)
        ,xlim=c(xlim_lowend,xlim_uppend)
        ,sub=paste("mean=",round(mean(variable_values),2)
        ,        "sum=",round(total_sum,2)
        ,        "std=",round(sd(variable_values),2)
        ,        "min=",round(min(variable_values),2)
        ,         "max=",round(max(variable_values),2))
        ,           freq=TRUE,col=barcol, ylab=yLabel,xlab=xLabel )
    }
    
    #add mean
    abline(v=mean(variable_values),col="blue",lwd=3)
    
    #add reference value lines
    if (!is.na(meta$referenvalue_type)) {
        if (meta$referenvalue_type=='value') { 
            if (!is.na(meta$referencevalue)) {
                abline(v=meta$referencevalue,col="blue",lty="dotted",lwd=2)
            }
        }
        if (meta$referenvalue_type=='range') {
            if (!is.na(meta$referencevalue) )  { 
                abline(v=meta$referencevalue,col="blue",lty="dotted",lwd=2)
            }
            if (!is.na(meta$referencevalue_high)){ 
                abline(v=meta$referencevalue_high,col="blue",lty="dotted",lwd=2)
            }
        }
    }
    
    #add alarm limits 
    if (!is.na(meta$alarm_lowerlimit))  {abline(v=meta$alarm_lowerlimit,col="red",lty="dotted",lwd=2)}
    if (!is.na(meta$alarm_upperlimit) )  {abline(v=meta$alarm_upperlimit,col="red",lty="dotted",lwd=2)}
    
    #add value limits 
    if (!is.na(meta$lowerlimit) ) {abline(v=meta$lowerlimit,col="black",lwd=1)}
    if (!is.na(meta$upperlimit) ) {abline(v=meta$upperlimit,col="black",lwd=1)}
    legend("topright",  
         c("Mean","Refvalue","Alarm","Limits" ), 
         lty=c(1,4,1,4), 
         col=c("blue","blue","red","grey"), 
         horiz=FALSE, 
         cex=0.8)
    return(h)
}

varids = "<varids>"
oiids = "<oiids>"
oitype = "<oitype>"
imagetype = "<imagetype>"
starttime = "<starttime>"
endtime = "<endtime>"
timeunit = "<timeunit>"
outputfile = "<outputfile>"
localResultFile = "<localResultFile>"
resultUrl = "<resultUrl>"

output_data <- tryCatch(
		{			
			suppressWarnings(
					{
					    if (imagetype == "vector") {
        					svglite(localResultFile);
    					} else {
							png(filename=localResultFile)
						}
						dbcon <- connectDB()
						data <- deploy_histogram(varids,oitype, oiids,starttime,endtime,timeunit,dbcon,imagetype);
						data["image"] = resultUrl
						data
					})
		},
		error=function(cond) {
			message(cond)
			return_list <- list(cond$message)
			names(return_list) <-c("error")
			return(return_list)
		},
		warning=function(cond) {
			message(cond)
			return(cond)
		},
		finally={
			dev.off()
		})
print(output_data)
	write(toJSON(output_data), file=outputfile)

File: src/main/resources/static/templates/visualizations/histogram.js
//OpenVA - Open software platform for visual analytics

//Copyright (c) 2018, VTT Technical Research Centre of Finland Ltd
//All rights reserved.
//Redistribution and use in source and binary forms, with or without
//modification, are permitted provided that the following conditions are met:

//1) Redistributions of source code must retain the above copyright
//notice, this list of conditions and the following disclaimer.
//2) Redistributions in binary form must reproduce the above copyright
//notice, this list of conditions and the following disclaimer in the
//documentation and/or other materials provided with the distribution.
//3) Neither the name of the VTT Technical Research Centre of Finland nor the
//names of its contributors may be used to endorse or promote products
//derived from this software without specific prior written permission.

//THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS'' AND ANY
//EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE LIABLE FOR ANY
//DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
//(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
//LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
//ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
//(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
// @author Pekka Siltanen

function histogramChart(responsedata,window,id,visualizationMethod) {
	if ("error" in responsedata) {
		return(responsedata);
	}
	var dataObject = trimHistogram(responsedata);	
//	return d3HistogramChart(dataObject,id,onClickHistogram,tooltipHistogram);
	return d3HistogramChart(dataObject,id,null,null);
	
}

function trimHistogram(data, title, labelTitle, valueTitle){

	var dataObject = new Object;
	dataObject.mainTitle = data.main_title;
	dataObject.xTitle = data.x_title;
	dataObject.yTitle = data.y_title;
	dataObject.subTitle = data.sub_title;
	dataObject.upperLimit = data.upperlimit;
	dataObject.lowerLimit = data.lowerlimit;
	dataObject.alarmUpperlimit = data.alarm_upperlimit;
	dataObject.alarmLowerLimit = data.alarm_lowerlimit;
	dataObject.mean = data.mean;
	dataObject.sum = data.sum;
	dataObject.std = data.std;
	dataObject.min = data.min;
	dataObject.max = data.max;	
	
	var values = [];
	var breaks = data.matrix.breaks;
	var counts = data.matrix.counts;

	for (var i = 0; i< counts.length; i++) {
		values.push({"count":counts[i], "breakStart":breaks[i],"breakEnd":breaks[i+1]})
	}
	dataObject["values"]= values;
	return dataObject;
}

function d3HistogramChart(dataObject,id,tooltipFunction, onClickFunction) {
	var canvasSize = getCanvasSize();	
	var margin = canvasSize.margin;
	var window = canvasSize.window;
    var w = window.width;
    var h = window.height;
	var legendMargin = canvasSize.legendMargin;
    
	var values = dataObject.values;	

	var x = d3.scaleLinear().range([0, w -(margin.left + margin.right)]);
	var y = d3.scaleLinear().range([h -(margin.top+ margin.bottom),0]);	

	var svg = addSvgRoot(id,w,h);

	var chart = svg.append("g")
	.attr('class', 'bar-chart')
	.attr("transform","translate(" + margin.left + "," + margin.top + ")");

	var tip = null;

	if (tooltipFunction != null) {
		tip = d3.tip()
		.attr('class', 'd3-tip')
		.offset([-10, 0])
		.html(function(d,i) {	    
			return tooltipFunction(dataObject,i);
		})
		svg.call(tip);
	}

	x.domain([values[0].breakStart, values[values.length-1].breakEnd]).nice(); 
	var max = d3.max(values, function(d) { return parseFloat(d.count); });
	y.domain([0,max]).nice(); 

	
	var rect = chart.selectAll(".bar-histogram")
	.data(values)
	 .enter()
	.append("rect")
	.attr("class", function(d) { return "bar-histogram"; })
	.attr("y", function(d) {return(y(d.count))})
	.attr("x", function(d) { return x(d.breakStart); })
	.attr("height", function(d) {return( y(0) - y(d.count));})
	.attr("width", function(d) { return x(d.breakEnd - d.breakStart);})

	if (onClickFunction != null) {
		rect.on('click', function(d,i) {
			onClickFunction(dataObject,i);
		});
	}    

	if (tooltipFunction != null) {    
		rect.on('mouseover', tip.show)
		.on('mouseout', tip.hide)
	}

	var xTranslation = y(0);
	var xAxis = d3.axisBottom(x);
	addXAxis(chart,xAxis,xTranslation, w- (margin.left + margin.right));
	addMainTitle(dataObject.mainTitle,svg,h,margin,0,"mainTitle");
    addXAxisTitle(chart, w, h, margin, dataObject.xTitle,"xAxisTitle");     
    addYAxisTitle(chart, w, h, margin, dataObject.yTitle); 
    addSubXTitle(chart, w, h, margin, dataObject.subTitle,"xAxisTitle");
    
    if (dataObject.mean != null) {
        addLine(chart,x,y,dataObject.mean,dataObject.mean,y.invert(0),y.invert(h -(margin.top+ margin.bottom)),"bar-meanline",null);
    }

    var legendText;
    for (var key in dataObject) {
        if (dataObject.hasOwnProperty(key)) {
				 switch (key) {
				 	case "alarmUpperLimit":
				 	case "alarmLowerlimit":
				 		legendText = "Alarm";
				 		break;
				 	case "upperLimit":
				 	case "lowerlimit":
				 		legendText = "Alarm";
				 		break;
				 	case "std":
				 		legendText = "Std";
				 		break;
				 	case "mean":	
				 		legendText = "Mean";
				 		break;
				 	case "min":
				 		legendText = "Min";
				 		break;
				 	case "max":	
				 		legendText = "Max";
				 		break;
				 	default:
				 		continue;
				 }
				 if (x(dataObject[key]) != null && x(dataObject[key]) > 0 && x(dataObject[key]) < w- (margin.left + margin.right)) {
					 addLine(chart,x,y,dataObject[key],dataObject[key],y.invert(0),y.invert(h -(margin.top+ margin.bottom)),"line-hist-" + key, legendText);
				 }
        }
    }
    

 
//	var textTranslation = 20; 
//
//	
	var yAxis = d3.axisLeft(y).ticks(10).tickFormat(d3.format("d"));
	chart.append("g")
	.attr("class", "y axis")
	.call(yAxis);
	
	var xTrans = w - legendMargin.right;
	var yTrans = legendMargin.top;
	var legend = svg.append("g")
	    .attr("class","legend")
	    .attr("transform","translate(" + xTrans +"," + yTrans +")")
	    .style("font-size","12px")
	    .call(d3.simpleLegend);
		
	return d3.select("#" + id + "> div").node();

}

function onClickHistogram(dataObject,i) {
	var message = "<div>" + dataObject.columnNames[i] + ": " + d3.format(".2f")(dataObject.amounts[i]) + "</div>";
	$(message).dialog();
}

function tooltipHistogram(dataObject,i) {
	return "<span style='color:red'>" + dataObject.columnNames[i] + ": " + d3.format(".2f")(dataObject.amounts[i]) + "</span>";
}
